### PR TITLE
feat: [PL-32568]: Parallelize ACL processing for User Group events.

### PR DESCRIPTION
**Pull request by Jimit09Gandhi**
## Describe your changes
Currently, the User Group is managed by NGManager and ACS listens to User Group CRUD events from Entity CRUD stream in Redis and maintains a copy of User Group in ACS. Now, in the new design, as part of processing of User Group events, ACLs would also be computed. This would help process user group events in parallel.
The change is per documented https://harness.atlassian.net/wiki/spaces/PLATFORM/pages/21361983548/Access+Control+Service+Design+Improvements#Parallelize-User-Group-Processing 

Tests:

- Add user to User Group  and acls are added.
- Remove user from User Group and acls are removed.
- Implicit ACLs Generated 

## Checklist
- [X] I've documented the changes in the PR description.
- [X] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49527)

### CODE REVIEW

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> Reviewed 18 of 19 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49527)* status: Some of the checks are too old. 19 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`access-control/config/config.yml` line 280 at r2](https://reviewable.io/reviews/harness/harness-core/49527#-NZjnXop0WbjjWza8vTx:-NZjnXop0WbjjWza8vTy:b-ie0b4h) ([raw file](https://github.com/harness/harness-core/blob/38adc567b0c70e522743e3776d34c798cc29115f/access-control/config/config.yml#L280)):*
> ```YAML
> disableRedundantACLs: ${DISABLE_REDUNDANT_ACLS:-false}
> 
> enableParallelProcessingOfACLsForUserGroups: ${ENABLE_PARALLEL_PROCESSING_OF_ACLS_FOR_USER_GROUPS:-false}
> ```

enableParallelProcessingOfUserGroupUpdates
___
*[`access-control/libs/aggregator/src/main/java/io/harness/aggregator/consumers/UserGroupChangeConsumer.java` line 67 at r2](https://reviewable.io/reviews/harness/harness-core/49527#-NZjvpq06MAWYY9E0HOR:-NZjvpq06MAWYY9E0HOS:b-ls37jp) ([raw file](https://github.com/harness/harness-core/blob/38adc567b0c70e522743e3776d34c798cc29115f/access-control/libs/aggregator/src/main/java/io/harness/aggregator/consumers/UserGroupChangeConsumer.java#L67)):*
> ```Java
> 
>   @Override
>   public boolean consumeUpdateEvent(String id, UserGroupUpdateEventData userGroupUpdateEventData) {
> ```

What are the differences in this consumer from already existing consumer?
___
*[`access-control/libs/aggregator/src/main/java/io/harness/aggregator/models/UserGroupUpdateEventData.java` line 24 at r2](https://reviewable.io/reviews/harness/harness-core/49527#-NZjpvTZDxGdtF3jIhYi:-NZjpvTZDxGdtF3jIhYj:b-wqzq63) ([raw file](https://github.com/harness/harness-core/blob/38adc567b0c70e522743e3776d34c798cc29115f/access-control/libs/aggregator/src/main/java/io/harness/aggregator/models/UserGroupUpdateEventData.java#L24)):*
> ```Java
> @OwnedBy(PL)
> public class UserGroupUpdateEventData implements AccessControlChangeEventData {
>   Set<String> userIdsAdded;
> ```

usersAdded
usersRemoved
___
*[`access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/HarnessUserGroupServiceImpl.java` line 72 at r2](https://reviewable.io/reviews/harness/harness-core/49527#-NZjsQ7c7ID7MhBUHYGl:-NZjsQ7c7ID7MhBUHYGm:b-7unpak) ([raw file](https://github.com/harness/harness-core/blob/38adc567b0c70e522743e3776d34c798cc29115f/access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/HarnessUserGroupServiceImpl.java#L72)):*
> ```Java
>           if (userGroupOptional.isPresent()) {
>             UserGroup existingUserGroup = userGroupOptional.get();
>             if (!userGroup.equals(existingUserGroup)) {
> ```

Are you sure they will ever be equal? Because 1 we are fetching from NG manager and another one from Access Control DB. lastModifiedAt will differ, isn't it?

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49527)* status: Some of the checks are too old. 19 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/HarnessUserGroupServiceImpl.java` line 72 at r2](https://reviewable.io/reviews/harness/harness-core/49527#-NZjsQ7c7ID7MhBUHYGl:-NZz-r2D5TUKE40sNAag:bj7lyc9) ([raw file](https://github.com/harness/harness-core/blob/f40f4bf97c3311a039e635d536a61961b81d033c/access-control/service/src/main/java/io/harness/accesscontrol/principals/usergroups/HarnessUserGroupServiceImpl.java#L72)):*
<details><summary><i>Previously, ankit-harness (Ankit Singhal) wrote…</i></summary><blockquote>

Are you sure they will ever be equal? Because 1 we are fetching from NG manager and another one from Access Control DB. lastModifiedAt will differ, isn't it?
</blockquote></details>

We are excluding createdAt, lastModifiedAt and version field for equality checks using 
@EqualsAndHashCode.Exclude. And performing on equality on fields which are meaningful. They can be equal because on NGManager side there may be update to usergroup's other fields like which ACS is not interested in like notificationConfigs, SSO config, etc.

**Review by ankit-harness — DISMISSED** [review]
:lgtm:

> Reviewed 8 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49527#-:-NZzeh655M4U0vqTCzi7:bnfp4nl)* status: 19 file(s) to review. 1 of 2 approvals obtained.

**Comment by Jimit09Gandhi** [comment]
trigger messagecheck

**Comment by Jimit09Gandhi** [comment]
trigger uts

**Comment by Jimit09Gandhi** [comment]
trigger uts

File: access-control/config/config.yml [hunk 1]
```diff
@@ -275,4 +275,6 @@ featureFlagConfig:
   featureFlagSystem: LOCAL
   syncFeaturesToCF: false
 
-disableRedundantACLs: ${DISABLE_REDUNDANT_ACLS:-false}
\ No newline at end of file
+disableRedundantACLs: ${DISABLE_REDUNDANT_ACLS:-false}
+
+enableParallelProcessingOfUserGroupUpdates: ${ENABLE_PARALLEL_PROCESSING_OF_USERGROUP_UPDATES:-false}
```
Comment (meenakshi-harness) [inline access-control/config/config.yml:280]: should we add these in replace configs?

**Review by Jimit09Gandhi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49527)* status: 20 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`access-control/config/config.yml` line 280 at r5](https://reviewable.io/reviews/harness/harness-core/49527#gh-1263385834:-N_IKqAp6pg5W43FE9bd:bx29coc) ([raw file](https://github.com/harness/harness-core/blob/d64c19845f9de2b3e15c904599cd5569a623bcb4/access-control/config/config.yml#L280)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

should we add these in replace configs?
</blockquote></details>

Not needed as these are read as env variables.

**Comment by Jimit09Gandhi** [comment]
trigger ut1

**Pull request merged**